folding-schemes/src/folding/circuits/cyclefold.rs

```diff
@@ -346,33 +357,64 @@ where
     for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
 {
     fn generate_constraints(self, cs: ConstraintSystemRef<CF2<C>>) -> Result<(), SynthesisError> {
-        let r_bits: Vec<Boolean<CF2<C>>> = Vec::new_witness(cs.clone(), || {
-            Ok(self.r_bits.unwrap_or(vec![false; N_BITS_RO]))
+        let r_bits: Vec<Vec<Boolean<CF2<C>>>> = self
+            .r_bits
+            // n_points-1, bcs is one for each point except for the first one
+            .unwrap_or(vec![vec![false; N_BITS_RO]; self.n_points - 1])
+            .iter()
+            .map(|r_bits_i| {
+                Vec::<Boolean<CF2<C>>>::new_witness(cs.clone(), || Ok(r_bits_i.clone()))
+            })
+            .collect::<Result<Vec<Vec<Boolean<CF2<C>>>>, SynthesisError>>()?;
+        let points = Vec::<GC>::new_witness(cs.clone(), || {
+            Ok(self.points.unwrap_or(vec![C::zero(); self.n_points]))
         })?;
-        let p1 = GC::new_witness(cs.clone(), || Ok(self.p1.unwrap_or(C::zero())))?;
-        let p2 = GC::new_witness(cs.clone(), || Ok(self.p2.unwrap_or(C::zero())))?;
-        // Fold the original Nova instances natively in CycleFold
-        // For the cmW we're computing: U_i1.cmW = U_i.cmW + r * u_i.cmW
-        // For the cmE we're computing: U_i1.cmE = U_i.cmE + r * cmT + r^2 * u_i.cmE, where u_i.cmE
+
+        #[cfg(test)]
+        {
+            assert_eq!(self.n_points, points.len());
+            assert_eq!(self.n_points - 1, r_bits.len());
+        }
+
+        // Fold the original points of the instances natively in CycleFold.
+        // In Nova,
+        // - for the cmW we're computing: U_i1.cmW = U_i.cmW + r * u_i.cmW
+        // - for the cmE we're computing: U_i1.cmE = U_i.cmE + r * cmT + r^2 * u_i.cmE, where u_i.cmE
         // is assumed to be 0, so, U_i1.cmE = U_i.cmE + r * cmT
-        let p3 = &p1 + p2.scalar_mul_le(r_bits.iter())?;
+        let mut p_folded: GC = points[0].clone();
+        // iter over n_points-1 because the first point is not multiplied by r^i (it is multiplied
+        // by r^0=1)
+        for i in 0..self.n_points - 1 {
+            p_folded += points[i + 1].scalar_mul_le(r_bits[i].iter())?;
+        }
```

Comment:
then in the `.scalar_mul_le(r_bits[i]` would be `.scalar_mul_le(r_bits[i-1]`. I think I prefer to have a `i+1` than a `i-1`, how do you see it?

Reply:
OOOOOH! I missed the second `i` ussage. This can be done easily with a peekable iterator IIUC. But feel free to leave it as is :)

Reply:
I'm checking the peekable iterator (didn't know it existed, thanks for discovering it to me ^^) and it seems that it would overcomplicate the code for something that we're already achieving with a simple line with `i` and `i+1` in
```rust
points[i+1].scalar_mul_le(r_bits[i].iter())?;
```
if it's ok, will leave it as it is.